proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,717 @@
+  !9 = !{i32 2, i32 1, i32 1, i32 0, i32 0, i32 -1}
+  !10 = !{i32 0, i32 8, i32 1, i32 0, i32 1, i32 -1}
+  !11 = !{i32 1, i32 -1, i32 1, i32 0, i32 1, i32 -1}
+
+  !12 = !{i32 0, i32 0, !13}
+  !13 = !{i32 3, i32 4, i32 0, i32 1, i32 0, i32 -1}
+
+  !14 = !{i32 1, i32 0, i32 10, i32 0, i32 3}
+
+  !15 = !{!16, !17}
+  !16 = !{i32 85, i32 1, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 1, i32 0, i32 0}
+  !17 = !{i32 21, i32 3, i32 1, i32 1, float 0.000000e+00, i32 16, i32 4, i32 2, float 0.000000e+00, float 0x47EFFFFFE0000000, i32 2, i32 0, i32 0}
+
+```
+Put the metadata into format table so it is easier to understand.
+```
+  Root Signature
+| | Version | Flag | RootParameters | StaticSamplers |
+|-| --- | --- | --- | --- |
+| | i32 | i32 | list of RootParameters (RootConstant, RootDescriptor, DescriptorTable) | list of StaticSamplers |
+| !3 | i32 2 | i32 3 | !4 | !15 |
+
+RootConstant
+| | ParameterType | Visibility | Register | Space | Num32BitValues |
+|-| --- | --- | --- | --- | --- |
+| | i32 |  i32 | i32 | i32 | i32 |
+| !14 | i32 1 |  i32 0 | i32 10 | i32 0 | i32 3 |
+
+RootDescriptor
+| | ParameterType | Visibility | Register | Space | Flags |
+|-| --- | --- | --- | --- | --- |
+| | i32 |  i32 | i32 | i32 | i32 |
+| !5 | i32 2 |  i32 0 | i32 0 | i32 1 | i32 8 |
+| !6 | i32 3 |  i32 0 | i32 0 | i32 0 | i32 0 |
+| !7 | i32 4 |  i32 0 | i32 0 | i32 0 | i32 0 |
+
+DescriptorTable
+| | ParameterType | Visibility | DescriptorRanges
+|-| --- | --- | --- |
+|  | i32 |  i32 | list of DescriptorRanges |
+| !8 | i32 0 |  i32 0 | !9, !10, !11 |
+| !12 | i32 0 |  i32 0 | !13 |
+
+DescriptorRange
+| | RangeType | NumDescriptors | BaseShaderRegister | Space | Flags | Offset |
+|-| --- | --- | --- | --- | --- | --- |
+| | i32 | i32 | i32 | i32 | i32 | i32 |
+| !9 | i32 2 | i32 1 | i32 1 | i32 0 | i32 0 | i32 -1 |
+| !10 | i32 0 | i32 8 | i32 1 | i32 0 | i32 1 | i32 -1 |
+| !11 | i32 1 | i32 -1 | i32 1 | i32 0 | i32 1 | i32 -1 |
+| !13 | i32 3 | i32 4 | i32 0 | i32 1 | i32 0 | i32 -1 |
+
+StaticSampler
+
+|| Filter | AddressU | AddressV | AddressW | MipLODBias | MaxAnisotropy | ComparisonFunc | BoarderColor | MinLOD | MaxLOD | Register | Space | Visibility |
+| - | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| | i32 | i32 | i32 | i32 | float | i32 | i32 | i32 | float | float | i32 | i32 | i32 |
+| !16 | i32 85 | i32 1 | i32 1 | i32 1 | float 0.000000e+00 | i32 16 | i32 4 | i32 2 | float 0.000000e+00 | float 0x47EFFFFFE0000000 | i32 1 | i32 0 | i32 0 |
+| !17 | i32 21 | i32 3 | i32 1 | i32 1 | float 0.000000e+00 | i32 16 | i32 4 | i32 2 | float 0.000000e+00 | float 0x47EFFFFFE0000000 | i32 2 | i32 0 | i32 0 |
+
+```
+
+
+## Alternatives considered (Optional)
+
+* Save root signature string to AST.
+
+  Assuming the cost to parse root signature string is cheap,
+  HLSLRootSignatureDecl could follow common Clang approach which only save 
```

Comment:
Following the common clang approach is a good starting point.  More investigation would be needed to really bottom out on how expensive string parsing is and how often it needs to be parsed.  If it is parsed once into an object that can be passed to other places to be inspected that might not be too bad, but I don't fully understand how clang deals with strings in the AST yet.  Maybe it stays as a string forever and travels around to consumers of the AST.